_posts/2022/2022-12-05-find-list-debug-active-record-callbacks-in-the-console.md

```diff
@@ -0,0 +1,89 @@
+
+## Use...
+
+...a loop in the Rails console.
+
+Open the console.
+
+```shell
+bin/rails console
+```
+
+Paste this code into the console to see the user-defined callbacks on a model:
+
+```ruby
+YourModel.__callbacks.each_with_object(Hash.new([])) do |(k, callbacks), result|
+  next if k == :validate # ignore validations
+  
+  callbacks.each do |c|
+    # remove autosaving callbacks from result
+    next if c.filter.to_s.include?("autosave")
+    next if c.filter.to_s.include?("_ensure_no_duplicate_errors")
+
+    result["#{c.kind}_#{c.name}"] += [c.filter]
+  end
+end
+```
+
+For the `User` object in my application, which uses [the Devise gem](https://github.com/heartcombo/devise) for authentication, the results look like:
+
+```ruby
+=> {
+  "before_validation"=>[:downcase_keys, :strip_whitespace, :set_timestamps_for_agreements],
+  "after_save"=>[#<Proc:0x00000001133b94d0 /.../gems/money-rails-1.15.0/lib/money-rails/active_record/monetizable.rb:148>],
+  "before_save"=>[:ensure_authentication_token],
+  "after_create"=>[:skip_reconfirmation_in_callback!],
+  "before_create"=>[:generate_confirmation_token],
+  "after_update"=>[:send_password_change_notification, :send_email_changed_notification],
+  "before_update"=>[:clear_reset_password_token :postpone_email_change_until_confirmation_and_regenerate_confirmation_token],
+  "after_commit"=>[:send_on_create_confirmation_instructions, :send_reconfirmation_instructions],
+}
+```
+
+Here you can see the callbacks inserted into the `User` model by our configuration of the `devise` gem. These aren't listed in the `app/models/user.rb` file.
+
+You can also see a dynamic `Proc`-based callback inserted by our use of the `money-rails` gem.
+
+
+## How does this work?
+
+The `#__callbacks` method on a model is [defined as a class method via Active Support](https://github.com/rails/rails/blob/main/activesupport/lib/active_support/callbacks.rb#L68). If accessed directly, like calling `YourModel.__callbacks`, it returns all the details of the logic performed by an Active Record instance before or after changing its state.
+
+The code above only pulls out the "headline" method calls
+
+In the specific code above I’ve presumed the following:
+
+You aren't interested in seeing all the validations: code like `validates_presence_of :attribute` creates a an entry in callback chain. We skip displaying these with `next if k == :validate`.
+
+You aren't interested in the callbacks that automatically save the values of associated models where you `have_many :related_models`. Hence `next if c.filter.to_s.include?("autosave")`.
+
+
+## Why?
+
+When you’re new to a complex application this may be useful as a debugging tool.
+
+I tend to limit my use of callbacks as much as possible due to the ease with which I‘ve confused myself before, but if you’re in a code base that uses them a lot, this might get you up to speed with the existing behaviour.
```

Comment:
*codebase?

Reply:
*..existing behaviour more quickly?

Reply:
```suggestion
I tend to limit my use of callbacks as much as possible due to the ease with which I‘ve confused myself before, but if you’re in a codebase that uses them a lot, this might get you up to speed with the existing behaviour more quickly.
```